flang/docs/DebugGeneration.md

```diff
@@ -0,0 +1,452 @@
+
+1. `DICompositeTypeAttr` will have a rank field which will be an expression.
+It will be used to get the rank value from descriptor.
+2. Instead of `DISubrangeType` for each dimension, there will be a single
+`DIGenericSubrange` which will allow debuggers to calculate bounds in any
+dimension.
+
+### Pointers and Allocatables
+The pointer and allocatable will be represented using `DICompositeTypeAttr`. It
+is quirk of DWARF that scalar allocatable or pointer variables will show up in
+the debug info as pointer to scalar while array pointer or allocatable
+variables show up as arrays. The behavior is same in gfortran and classic flang.
+
+```
+  integer, allocatable :: ar(:)
+  integer, pointer :: sc
+
+!1 = !DILocalVariable(name: "sc", type: !2)
+!2 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !3, associated: !9 ...)
+!3 = !DIBasicType(tag: DW_TAG_base_type, name: "integer", ...)
+!4 = !DILocalVariable(name: "ar", type: !5 ...)
+!5 = !DICompositeType(tag: DW_TAG_array_type, baseType: !3, elements: !6, dataLocation: !8, allocated: !9)
+!6 = !{!7}
+!7 = !DISubrange(lowerBound: !10, upperBound: !11 ...)
+!8 = !DIExpression(DW_OP_push_object_address, DW_OP_deref)
+!9 = !DIExpression(DW_OP_push_object_address, DW_OP_deref, DW_OP_lit0, DW_OP_ne)
+!10 = !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 24, DW_OP_deref)
+!11 = !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 32, DW_OP_deref)
+
+```
+
+IN FIR, these variable are represent as <!fir.box<!fir.heap<>> or
+fir.box<!fir.ptr<>>. There is also `allocatable` or `pointer` attribute on
+the `DeclareOp`. This allows us to generate allocated/associated status of
+these variables. The metadata to get the information from the descriptor is
+similar to arrays.
+
+### Strings
+
+Fixed sized strings will be treated like fixed sizes arrays and
+allocatable strings will be treated like allocatable arrays. Metadata
+will be generated to enable debuggers to find their size.
+
+```
+  character(len=:), allocatable :: var
+  character(len=20) :: fixed
+
+!1 = !DILocalVariable(name: "var", type: !2)
+!2 = !DICompositeType(tag: DW_TAG_array_type, baseType: !3, elements: !4, dataLocation: !6)
+!3 = !DIBasicType(name: "char", size: 8, encoding: DW_ATE_unsigned_char)
+!4 = !DISubrange(count: !5, lowerBound: 1)
+!5 = !DIExpression(DW_OP_push_object_address, DW_OP_plus_uconst, 8, DW_OP_deref)
+!6 = !DIExpression(DW_OP_push_object_address, DW_OP_deref)
+
+!7 = !DILocalVariable(name: "fixed", type: !8)
+!8 = !DICompositeType(tag: DW_TAG_array_type, baseType: !3, size: 160, elements: !9)
+!9 = !DISubrange(count: 20, lowerBound: 1)
+```
+
+### Association
+
+They will be treated like normal variables. Although we may require to handle
+the case where the `DeclareOp` of one variable points to the `DeclareOp` of
+another variable (e.g. a => b).
+
+### Namelists
+
+FIR does not seem to have a way to extract information about namelists.
```

Comment:
Namelists are handled by creating [runtime data structures](https://github.com/llvm/llvm-project/blob/cb9cf331fa69bb8319886cdca043a742b4d19ce5/flang/runtime/namelist.h#L30) that describes them. These data structures are globals when possible, and locals otherwise (without fir.declare currently). They can be identified via [their mangling](https://github.com/llvm/llvm-project/blob/main/flang/docs/BijectiveInternalNameUniquing.md#namelist-groups) using [InternalNames.h](https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Support/InternalNames.h).

It may be non trivial to generate the debug info from the global/local. Maybe we could abstract that more in FIR/Lowering.

Reply:
Thanks for the information. Although when I tried a simple namelist example, I did not see a local or global with this name mangling scheme in fir. Ideally we will need something that represent namelist and then refer to `DeclareOp` of all the variables that are part of the given namelist.

Reply:
These data structures are created lazily when encountering the namelist in an IO statement, so if your test had no IO statements using it, you would indeed not have seen it. I understand this is not convenient from a debug info point of view, again more abstraction might be needed in lowering here to generate the debug info.